File: packages/tickle_core/lib/src/models/counter.dart
import 'package:equatable/equatable.dart';

class Counter extends Equatable {
  final String id;
  final String title;
  final String? emoji;
  final String colorHex;
  final int currentCount;
  final int? goalValue;
  final bool isArchived;
  final DateTime createdAt;
  final int sortOrder;

  const Counter({
    required this.id,
    required this.title,
    this.emoji,
    required this.colorHex,
    this.currentCount = 0,
    this.goalValue,
    this.isArchived = false,
    required this.createdAt,
    this.sortOrder = 0,
  });

  Counter copyWith({
    String? id,
    String? title,
    String? emoji,
    String? colorHex,
    int? currentCount,
    int? goalValue,
    bool? isArchived,
    DateTime? createdAt,
    int? sortOrder,
  }) {
    return Counter(
      id: id ?? this.id,
      title: title ?? this.title,
      emoji: emoji ?? this.emoji,
      colorHex: colorHex ?? this.colorHex,
      currentCount: currentCount ?? this.currentCount,
      goalValue: goalValue ?? this.goalValue,
      isArchived: isArchived ?? this.isArchived,
      createdAt: createdAt ?? this.createdAt,
      sortOrder: sortOrder ?? this.sortOrder,
    );
  }

  @override
  List<Object?> get props => [
        id,
        title,
        emoji,
        colorHex,
        currentCount,
        goalValue,
        isArchived,
        createdAt,
        sortOrder,
      ];
}

File: apps/tickle_ios/Tickle/Models/Counter.swift
import Foundation
import SwiftData

@Model
final class Counter {
    var id: String = UUID().uuidString
    var title: String = ""
    var emoji: String?
    var colorHex: String = "#3498DB"
    var currentCount: Int = 0
    var goalValue: Int?
    var isArchived: Bool = false
    var createdAt: Date = Date()
    var sortOrder: Int = 0
    var reminderHour: Int?
    var reminderMinute: Int?
    
    @Relationship(deleteRule: .cascade, inverse: \CounterLog.counter)
    var logs: [CounterLog]?
    
    init(id: String = UUID().uuidString,
         title: String,
         emoji: String? = nil,
         colorHex: String,
         currentCount: Int = 0,
         goalValue: Int? = nil,
         isArchived: Bool = false,
         createdAt: Date = Date(),
         sortOrder: Int = 0,
         reminderHour: Int? = nil,
         reminderMinute: Int? = nil) {
        self.id = id
        self.title = title
        self.emoji = emoji
        self.colorHex = colorHex
        self.currentCount = currentCount
        self.goalValue = goalValue
        self.isArchived = isArchived
        self.createdAt = createdAt
        self.sortOrder = sortOrder
        self.reminderHour = reminderHour
        self.reminderMinute = reminderMinute
        self.logs = []
    }
}
